tests/MongoDB.Driver.Tests/AsyncCursorTests.cs

```diff
@@ -31,6 +33,77 @@ namespace MongoDB.Driver.Tests
     public class AsyncCursorTests
     {
         //public methods
+        [SkippableTheory]
+        [ParameterAttributeData]
+        public void Cursor_should_call_close_in_thread_safe_way(
+            [Values(false, true)] bool cappedCollection,
+            [Values(false, true)] bool async)
+        {
+            string testCollectionName = "test";
+            string testDatabaseName = "test";
+            var client = CreateClient();
+            DropCollection(client, testDatabaseName, testCollectionName);
+            var database = client.GetDatabase(testDatabaseName);
+
+            var createCollectionOptions = new CreateCollectionOptions();
+            if (cappedCollection)
+            {
+                createCollectionOptions.Capped = true;
+                createCollectionOptions.MaxSize = 1000;
+            }
+
+            database.CreateCollection(testCollectionName, createCollectionOptions);
+            var collection = database.GetCollection<BsonDocument>(testCollectionName);
+            collection.InsertOne(new BsonDocument());
+
+            var findOptions = new FindOptions()
+            {
+                BatchSize = 1
+            };
+            if (cappedCollection)
+            {
+                findOptions.CursorType = CursorType.TailableAwait;
+            }
+            var findResult = collection.Find(FilterDefinition<BsonDocument>.Empty, findOptions);
+            IAsyncCursor<BsonDocument> cursor;
+            if (async)
+            {
+                cursor = findResult
+                    .ToCursorAsync()
+                    .GetAwaiter()
+                    .GetResult();
+            }
+            else
+            {
+                cursor = findResult.ToCursor();
+            }
+
+            var asyncCursor = cursor.As<AsyncCursor<BsonDocument>>();
+            var enumerable = asyncCursor
+                .ToEnumerable()
+                .GetEnumerator();
+
+            enumerable.MoveNext();
+
+            var thread = new Thread(
+                () =>
+                {
+                    Thread.Sleep(1000);
+                    asyncCursor.Close();
+                });
+            thread.Start();
+            var exception = Record.Exception(() => enumerable.MoveNext());
+
+            if (cappedCollection)
+            {
+                exception.Should().BeOfType<ObjectDisposedException>();
```

Comment:
In response to your question in your PR, I feel like this is a reasonable exception to see when calling MoveNext on a closed cursor.